csharp/src/Google.Protobuf/WritingPrimitives.cs

```diff
@@ -171,9 +179,36 @@ public static void WriteString(ref Span<byte> buffer, ref WriterInternalState st
             {
                 if (length == value.Length) // Must be all ASCII...
                 {
-                    for (int i = 0; i < length; i++)
+                    // If 64bit, and value has at least 4 chars, process 4 chars at a time.
+                    if (IntPtr.Size == 8 && value.Length >= 4)
+                    {
+                        ref byte sourceBytes = ref Unsafe.As<char, byte>(ref MemoryMarshal.GetReference(value.AsSpan()));
+                        ref byte destinationBytes = ref MemoryMarshal.GetReference(buffer.Slice(state.position));
+
+                        // Process 4 chars at a time until there are less than 4 remaining.
+                        // We already know all characters are ASCII so there is no need to validate the source.
+                        int currentIndex = 0;
+                        int lastIndexWhereCanReadFourChars = (2 * value.Length - 8);
+                        do
+                        {
+                            NarrowFourUtf16CharsToAsciiAndWriteToBuffer(
+                                ref Unsafe.AddByteOffset(ref destinationBytes, (IntPtr)(currentIndex / 2)),
+                                Unsafe.ReadUnaligned<ulong>(ref Unsafe.AddByteOffset(ref sourceBytes, (IntPtr)currentIndex)));
+
+                        } while ((currentIndex += 8) <= lastIndexWhereCanReadFourChars);
+
+                        // Process remainder.
```

Comment:
Maybe merge the remainder part with the else path below by setting the start index here?

Reply:
I updated it to use refs to Unsafe to set remainder in this loop.